trunner/runners/flasher.py

```diff
@@ -0,0 +1,108 @@
+#
+# Phoenix-RTOS test runner
+#
+# Flasher classes for device runners
+#
+# Copyright 2021, 2022 Phoenix Systems
+# Authors: Jakub Sarzyński, Damian Loewnau
+#
+
+from abc import ABC, abstractmethod
+import logging
+import sys
+
+from pexpect.exceptions import TIMEOUT, EOF
+from .common import PloError, PloTalker, RebootError
+from .wrappers import Psu, Gdb, Phoenixd, PhoenixdError, GdbError, append_output
+
+
+class Flasher(ABC):
+    """Common interface for flashing the system image"""
+
+    def __init__(self, target, serial_port, phoenixd_port, flash_memory, runner_reboot, logpath):
+        self.target = target
+        self.serial_port = serial_port
+        self.phoenixd_port = phoenixd_port
+        self.flash_memory = flash_memory
+        self.runner_reboot = runner_reboot
+        self.logpath = logpath
+
+    def flash(self):
+        phd = None
+        try:
+            self.reboot(flashing_mode=True)
+            with PloTalker(self.serial_port) as plo:
+                if self.logpath:
+                    plo.plo.logfile = open(self.logpath, "a")
+
+                self.upload_plo(plo)
+
+                with Phoenixd(self.target, self.phoenixd_port) as phd:
+                    plo.copy_file2mem(
+                        src='usb0',
+                        file='phoenix.disk',
+                        dst=f'flash{self.flash_memory}',
+                    )
+
+            self.reboot(flashing_mode=False)
+        except (TIMEOUT, EOF, PloError, PhoenixdError, GdbError, RebootError) as exc:
+            exception = f'{exc}\n'
+
+            if phd and not isinstance(exc, RebootError):
+                exception = append_output('phoenixd', exception, phd.output())
+
+            logging.info(exception)
+            sys.exit(1)
+
+    @abstractmethod
+    def reboot(self):
+        """Method used for setting the intended boot mode and restarting/resseting the board."""
+        pass
+
+    @abstractmethod
+    def upload_plo(self):
+        """Method used for flashing a device with the image containing tests."""
+        pass
+
+
+class ZYNQ7000JtagFlasher(Flasher):
+    """ Class intended for flashing the system image to Zynq7000 platform using jtag and plo"""
+
+    PLO_FILE = 'plo-gdb.elf'
+    GDB_SCRIPT = '/opt/phoenix/scripts/gdb_load_file'
+
+    def __init__(self, target, serial_port, phoenixd_port, flash_memory, runner_reboot, logpath):
+        self.long_flashing = True
+        super().__init__(target, serial_port, phoenixd_port, flash_memory, runner_reboot, logpath)
+
+    def reboot(self, flashing_mode):
+        if flashing_mode:
+            self.runner_reboot(jtag_mode=True)
+        else:
+            self.runner_reboot()
```

Comment:
I think runner_reboot should be passed directly to the `flash` function as a callback if we want to use it this way. Now it is a little bit confusing and hard to follow. We could make the common interface for the reboot function with flashing_mode argument or wrap it before calling flash method.

Reply:
There is my proposal, please confirm, whether it's ok:
```C
    def reboot(self, flashing_mode=False):
        """Reboots a tested device."""
        jtag_mode = False
        if flashing_mode is True:
            jtag_mode = True

        if self.is_rpi_host:
            self.rpi_reboot(jtag_mode)
        else:
            hostpc_reboot(jtag_mode)
```